pkgs/build-support/rust/build-rust-package/default.nix

```
@@ -15,145 +15,149 @@
+      else target;
+    sysroot = callPackage ./sysroot { } {
+      inherit target shortTarget;
+      RUSTFLAGS = finalAttrs.RUSTFLAGS or "";
+      originalCargoToml = finalAttrs.src + /Cargo.toml; # profile info is later extracted
+    };
+
+    cargoDeps =
+      if cargoVendorDir != null then null
+      else if cargoLock != null then importCargoLock cargoLock
+      else
+        fetchCargoTarball ({
+          src = finalAttrs.src or "";
+          srcs = finalAttrs.srcs or null;
+          sourceRoot = finalAttrs.sourceRoot or null;
+          unpackPhase = finalAttrs.unpackPhase or null;
+          cargoUpdateHook = finalAttrs.cargoUpdateHook or args.cargoUpdateHook or "";
+          # Name for the vendored dependencies tarball
+          name = finalAttrs.cargoDepsName or finalAttrs.name or "${finalAttrs.pname}-${finalAttrs.version}";
+          # see comment below
+          #patches = finalAttrs.cargoPatches;
+          patches = finalAttrs.cargoPatches or [ ];
+        } // lib.optionalAttrs (finalAttrs ? cargoHash) {
+          hash = finalAttrs.cargoHash;
+        } // lib.optionalAttrs (finalAttrs ? cargoSha256) {
+          sha256 = finalAttrs.cargoSha256;
+        } // (args.depsExtraArgs or { }));
+
+    maybeSetStr = x: lib.optionalString (previousAttrs ? ${x}) previousAttrs.${x};
+    maybeSetListFromPrevious = x: lib.optionals (previousAttrs ? ${x}) previousAttrs.${x};
+
+    # cargoBuildFeatures example:
+    # previous: use directly specified cargoBuildFeatures from the original derivation (the one in nixpkgs)
+    # final: use buildType from either the original derivation or one overriden with overrideAttrs
+    # default: use the default passed if none of the above are set
+    pickPreviousOrFinalOrDefault = previous: final: default: (previousAttrs.${previous} or finalAttrs.${final} or default);
+
+  in
+
+  # Tests don't currently work for `no_std`, and all custom sysroots are currently built without `std`.
+    # See https://os.phil-opp.com/testing/ for more information.
+  assert useSysroot -> !(finalAttrs.doCheck or true);
+  {
+    inherit buildAndTestSubdir cargoDeps;
+
+    # buildType is checked for 4 of the cargo built in types
+    # directly specifying cargoBuildType can be used as a escape-hatch
+    cargoBuildType = assert finalAttrs ? buildType -> lib.assertOneOf "buildType" finalAttrs.buildType [ "release" "debug" "test" "bench" ];
+      pickPreviousOrFinalOrDefault "cargoBuildType" "buildType" "release";
+
+    cargoCheckType = pickPreviousOrFinalOrDefault "cargoCheckType" "checkType" finalAttrs.cargoBuildType;
+
+    cargoBuildNoDefaultFeatures = pickPreviousOrFinalOrDefault "cargoBuildNoDefaultFeatures" "buildNoDefaultFeatures" false;
+
+    cargoCheckNoDefaultFeatures = pickPreviousOrFinalOrDefault "cargoCheckNoDefaultFeatures" "checkNoDefaultFeatures"
+      finalAttrs.cargoBuildNoDefaultFeatures;
+
+    cargoBuildFeatures = pickPreviousOrFinalOrDefault "cargoBuildFeatures" "buildFeatures" "";
```

```suggestion
    cargoBuildFeatures = pickPreviousOrFinalOrDefault "cargoBuildFeatures" "buildFeatures" [];
```
old default was `[]`, probably better to keep it like that?